src/github.com/getlantern/enproxy/conn_impl_reads.go

```diff
@@ -107,39 +92,14 @@ func (c *Conn) processReads() {
 
 // submitRead submits a read to the processReads goroutine, returning true if
 // the read was accepted or false if reads are no longer being accepted
-func (c *Conn) submitRead(b []byte) bool {
-	c.readMutex.RLock()
-	defer c.readMutex.RUnlock()
-	if c.doneReading {
+func (c *conn) submitRead(b []byte) bool {
+	c.closingMutex.RLock()
```

Comment:
We used to lock on the closed status.  It turns out that this could cause deadlocks.  In practice, we don't need to know if closing has completed, we just need to know that we're closing.